### PR TITLE
(feat): Add memfd secret based allocation

### PR DESCRIPTION
**Pull request by prabhpreet**
Adds functions to allocate and free memfd_secret based allocations on Linux only using `memfd_secret`, `memfd_secret_sized` and `free_memfd_secret` methods in `alloc` module, adding guard pages and canary padding.

Addresses #13 partly by adding functions to allocate using this method.

Plan to use for rosenpass/rosenpass#320

Starting this as a draft PR- please do suggest any comments/improvements. I'll be adding tests soon as well.

### CODE REVIEW

File: src/alloc.rs [hunk 1]
```diff
@@ -64,6 +66,41 @@ mod raw_alloc {
...
+
+    #[inline]
+    pub unsafe fn alloc_memfd_secret(size: usize) -> Option<NonNull<u8>> {
+        let fd: Result<libc::c_int, _> = libc::syscall(libc::SYS_memfd_secret, 0).try_into();
+
+        if fd.is_err() || fd.unwrap() < 0 {
+            return None;
+        }
+
+        let fd = fd.unwrap();
```
Comment (quininer) [inline src/alloc.rs:83]: we can use match or ? to avoid unwrap

**Comment by quininer** [comment]
This looks good to me, maybe should be put into a separate file and some tests need to be added.

File: memsec-test/Cargo.toml [hunk 2]
```diff
@@ -18,7 +18,8 @@ libsodium-sys = { version = "0.2" }
 nix = "0.26"
 
 [features]
-default = [ "alloc", "use_os" ]
+default = [ "alloc", "use_os", "alloc_ext"]
 nightly = [ "memsec/nightly" ]
 use_os = [ "memsec/use_os" ]
 alloc = [ "memsec/alloc" ]
+alloc_ext = [ "memsec/alloc_ext" ]
```
Comment (quininer) [inline memsec-test/Cargo.toml:25]: Should it contain use_os?

File: src/alloc/allocext.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    _memfd_secret(size).map(|memptr| {
+        ptr::write_bytes(memptr, GARBAGE_VALUE, size);
+        NonNull::new_unchecked(slice::from_raw_parts_mut(memptr, size))
+    })
+}
+
+/// Secure `free` for memfd_secret allocations,
+/// i.e. provides read write access back to mprotect guard pages
+/// and unmaps mmaped secrets
+#[cfg(target_os = "linux")]
```
Comment (quininer) [inline src/alloc/allocext.rs:113]: This cfg can be placed in mod top.

File: memsec-test/Cargo.toml (hunk 2 above)
Comment (prabhpreet) [reply memsec-test/Cargo.toml]: added it

File: src/alloc/allocext.rs (hunk 3 above)
Comment (prabhpreet) [reply src/alloc/allocext.rs]: moved it to a toplevel cfg for linux

File: src/alloc/allocext/linux.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    #[inline]
+    pub unsafe fn alloc_memfd_secret(size: usize) -> Option<NonNull<u8>> {
+        let fd: Result<libc::c_int, _> = libc::syscall(libc::SYS_memfd_secret, 0).try_into();
+
+        if fd.is_err() || fd.unwrap() < 0 {
+            return None;
+        }
+
+        let Ok(fd) = fd else {
```
Comment (quininer) [inline src/alloc/allocext/linux.rs:23]: This should be merge with 19L

File: src/alloc/allocext/linux.rs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+mod memfd_secret_alloc {
+    use core::convert::TryInto;
+
+    use super::*;
+
+    #[inline]
+    pub unsafe fn alloc_memfd_secret(size: usize) -> Option<NonNull<u8>> {
+        let fd: Result<libc::c_int, _> = libc::syscall(libc::SYS_memfd_secret, 0).try_into();
+
+        if fd.is_err() || fd.unwrap() < 0 {
```
Comment (quininer) [inline src/alloc/allocext/linux.rs:19]:
```suggestion
        if fd.is_err() || fd.unwrap() < 0 {
        let fd = fd.ok().filter(|fd| fd >= 0)?;
```

Comment (prabhpreet) [reply src/alloc/allocext/linux.rs]: Thanks, added! :)

**Comment by prabhpreet** [comment]
@quininer : Thanks for your review comments. I think this is ready for review now. 

I have also added extra tests as a sanity check to attempt to write in guard pages, and canary area to see if allocation fails at right stages.

File: src/alloc/allocext/linux.rs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    // check
+    if !crate::memeq(canary_ptr as *const u8, CANARY.as_ptr(), CANARY_SIZE) {
+        abort();
+    }
+
+    // free
+    let total_size = PAGE_SIZE + PAGE_SIZE + unprotected_size + PAGE_SIZE;
+    _mprotect(base_ptr, total_size, Prot::ReadWrite);
+
```
Comment (quininer) [inline src/alloc/allocext/linux.rs:121]: Should there be a memzero here?

File: src/alloc/allocext/linux.rs [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+        return None;
+    }
+
+    // aligned alloc ptr
+    let size_with_canary = CANARY_SIZE + size;
+    let unprotected_size = page_round(size_with_canary);
+    let total_size = PAGE_SIZE + PAGE_SIZE + unprotected_size + PAGE_SIZE;
+    let (base_ptr, fd) = alloc_memfd_secret(total_size)?;
+    let base_ptr = base_ptr.as_ptr();
+    let fd_ptr = base_ptr.add(size_of::<usize>());
```
Comment (quininer) [inline src/alloc/allocext/linux.rs:55]: maybe we should add an assert size + fd <= page

**Review by quininer — CHANGES_REQUESTED** [review]

File: src/alloc/allocext/linux.rs (hunk 7 above)
Comment (prabhpreet) [reply src/alloc/allocext/linux.rs:56]: Thank you, made the change on line 43. Please let me know if that wasn't what you had in mind.

File: memsec-test/tests/allocext_linux.rs [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq)]
+enum TestState {
+    Init,
+    Allocate,
+    Operation,
+    Free,
+}
+
+/// Attempts to
+#[cfg(unix)]
```
[the remark this answers is not in the document]
Comment (prabhpreet) [reply memsec-test/tests/allocext_linux.rs]: Removed

File: src/alloc/allocext/linux.rs (hunk 6 above)
Comment (prabhpreet) [reply src/alloc/allocext/linux.rs:122]: Thank you, added it with crate::memzero

Comment (prabhpreet) [reply src/alloc/allocext/linux.rs:122]: Added it for entire memory length, starting with base_ptr

Comment (quininer) [reply src/alloc/allocext/linux.rs:122]: I think unprotected_ptr is enough, like https://github.com/quininer/memsec/blob/master/src/alloc.rs#L223C20-L223C35

Comment (prabhpreet) [reply src/alloc/allocext/linux.rs:122]: Done

File: memsec-test/tests/malloc.rs [hunk 9]
```diff
@@ -83,10 +83,12 @@ fn malloc_mprotect_1_test() {
     }
 }
 
-procspawn::enable_test_support!();
+#[cfg(unix)]
```
Comment (quininer) [inline memsec-test/tests/malloc.rs:86]: You can put it into a separate file.

**Comment by prabhpreet** [comment]
@quininer Looks like the checks have passed. When you think it is ready, I request you to please squash the commit changes before merging.

**Comment by quininer** [comment]
Thank you!

**Pull request merged**